### PR TITLE
refactor(subscriptions): skip `bullmq` for WS subs, push all WS subs thru one Redis channel

### PR DESCRIPTION
**Pull request by ThatOneBro**

### CODE REVIEW

File: packages/server/src/subscriptions/websockets.ts [hunk 1]
```diff
@@ -22,12 +22,86 @@ export interface BindWithTokenMsg extends BaseSubscriptionClientMsg {
...
+    heartbeatHandler = (): void => {
+      for (const [ws, subscriptionIds] of wsToSubLookup.entries()) {
+        ws.send(JSON.stringify(createSubHeartbeatEvent(subscriptionIds)));
+      }
+    };
+    heartbeat.addEventListener('heartbeat', heartbeatHandler);
+  }
+}
+
+// function unsubscribeWsFromSubscription(ws: ws.WebSocket, subscriptionId: string): void {
```
Comment (codyebberson) [inline packages/server/src/subscriptions/websockets.ts:71]: Cleanup

File: packages/server/src/subscriptions/websockets.ts [hunk 2]
```diff
@@ -22,12 +22,86 @@ export interface BindWithTokenMsg extends BaseSubscriptionClientMsg {
...
+      for (const socket of subToWsLookup.get(subscriptionId) ?? []) {
+        socket.send(JSON.stringify(bundle), { binary: false });
+      }
+    }
+  });
+  await redisSubscriber.subscribe('medplum:subscriptions:r4:websockets');
+}
+
+function subscribeWsToSubscription(ws: ws.WebSocket, subscriptionId: string): void {
+  let wsSet = subToWsLookup.get(subscriptionId);
```
Comment (codyebberson) [inline packages/server/src/subscriptions/websockets.ts:46]: 🤔  Is it allowed to use the same `Subscription` on multiple websocket connections? 🤔

File: packages/server/src/subscriptions/websockets.ts [hunk 3]
```diff
@@ -22,12 +22,86 @@ export interface BindWithTokenMsg extends BaseSubscriptionClientMsg {
...
+    return;
+  }
+  for (const subscriptionId of subscriptionIds) {
+    if (!subToWsLookup.has(subscriptionId)) {
+      globalLogger.error(`[WS] Subscription binding to subscription ${subscriptionId} for this WebSocket is missing`);
+      continue;
+    }
+    const wsSet = subToWsLookup.get(subscriptionId) as Set<ws.WebSocket>;
+    wsSet.delete(ws);
+    if (wsSet.size === 0) {
```
Comment (codyebberson) [inline packages/server/src/subscriptions/websockets.ts:95]: Clever

**Review by codyebberson — APPROVED** [review]
Nice, this is clean.

1. Minor, some commented out code
2. What is the case for multiple websocket connections sharing a `Subscription`?  I guess it's neat that we support it, I just didn't think it was something that would be possible.  Maybe if the Subscription id is in localStorage, and you have multiple tabs open?

File: packages/server/src/subscriptions/websockets.ts (hunk 2 above)
Comment (ThatOneBro) [reply packages/server/src/subscriptions/websockets.ts:46]: It's technically allowed by the spec, and  Currently our client doesn't support it but it shouldn't be hard to add support for it in the future. I kinda left it open more for other client implementers / to be more FHIR spec-compliant

**Pull request merged**